toolkit/tools/internal/grub/grubtokenizer.go

```diff
@@ -0,0 +1,509 @@
+	return nil
+}
+
+// Parses single quoted string.
+// Single quoted strings don't accept any character escapes and end at the first ' character.
+func (t *grubConfigTokenizer) parseSingleQuotedString() error {
+	locStart := t.scanner.Loc()
+
+	// Consume ' char.
+	t.scanner.Next()
+
+	sb := strings.Builder{}
+loop:
+	for {
+		c, eof := t.scanner.Peek()
+		if eof {
+			return fmt.Errorf("unexpected end-of-file during single-quoted string (%d:%d)", t.scanner.Line(),
+				t.scanner.Col())
+		}
+
+		switch c {
+		// End of string
+		case '\'':
+			// Consume ' char.
+			t.scanner.Next()
+			break loop
+
+		// Normal character
+		default:
+			sb.WriteRune(c)
+			t.scanner.Next()
+		}
+	}
+
+	subWord := t.newSubWord(locStart, t.scanner.Loc(), STRING, sb.String())
+	t.subWords = append(t.subWords, subWord)
+	return nil
+}
+
+// Parses variable expansions.
+// e.g. $a, ${a}
+func (t *grubConfigTokenizer) parseVariableExpansion(subWordType SubWordType) error {
+	locStart := t.scanner.Loc()
+
+	// Consume $ char.
+	t.scanner.Next()
+
+	// Check if name is surrounded by braces.
+	c, eof := t.scanner.Peek()
+	usesBraces := false
+	if !eof && c == '{' {
+		usesBraces = true
+		// Consume { char
+		t.scanner.Next()
+	}
+
+	sb := strings.Builder{}
+loop:
+	for {
+		c, eof := t.scanner.Peek()
+		if eof {
+			break loop
+		}
+
+		switch {
+		// Name characters.
+		case (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '_':
+			sb.WriteRune(c)
+			t.scanner.Next()
+
+		default:
+			break loop
+		}
+	}
+
+	varName := sb.String()
+	if usesBraces {
+		c, eof := t.scanner.Peek()
+		if eof {
+			return fmt.Errorf("unexpected end-of-file while parsing variable expansion (%d:%d)", t.scanner.Line(),
+				t.scanner.Col())
+		}
+
+		switch c {
+		case '}':
+			// Consume } char
+			t.scanner.Next()
+
+		default:
+			return fmt.Errorf("missing } in variable expansion (%d:%d)", t.scanner.Line(), t.scanner.Col())
+		}
+
+		if varName == "" {
+			return fmt.Errorf("variable expansion missing name (%d:%d)", locStart.Line, locStart.Col)
+		}
+	} else if varName == "" {
+		// Name is invalid. So, $ is interpreted as a normal character.
+		subWord := t.newSubWord(locStart, t.scanner.Loc(), STRING, "$")
+		t.subWords = append(t.subWords, subWord)
+		return nil
+	}
+
+	subWord := t.newSubWord(locStart, t.scanner.Loc(), subWordType, varName)
+	t.subWords = append(t.subWords, subWord)
+	return nil
+}
+
+func (t *grubConfigTokenizer) newToken(locStart, locEnd filescanner.SourceLoc, tokenType TokenType) Token {
+	rawContentStart := locStart.Index
+	rawContentEnd := locEnd.Index
+	rawContent := t.scanner.Content()[rawContentStart:rawContentEnd]
+
+	token := Token{
+		Loc: SourceLoc{
+			Start: locStart,
+			End:   locEnd,
+		},
+		Type:       tokenType,
+		RawContent: rawContent,
+	}
+	return token
+}
+
+func (t *grubConfigTokenizer) newSubWord(locStart, locEnd filescanner.SourceLoc, subWordType SubWordType, value string,
```

Comment:
would it make sense to document/define what a `token`, a `subword`, and a `word` are?

Reply:
Added some code comments on the types. Please let me know if you think they need more details.